skrifa/src/meta/metrics.rs

```diff
@@ -0,0 +1,434 @@
+    /// Number of glyphs in the font.
+    pub glyph_count: u16,
+    /// True if the font is not proportionally spaced.
+    pub is_monospace: bool,
+    /// Italic angle in counter-clockwise degrees from the vertical. Zero for upright text,
+    /// negative for text that leans to the right.
+    pub italic_angle: f32,
+    /// Distance from the baseline to the top of the alignment box.
+    pub ascent: f32,
+    /// Distance from the baseline to the bottom of the alignment box.
+    pub descent: f32,
+    /// Recommended additional spacing between lines.
+    pub leading: f32,
+    /// Distance from the baseline to the top of a typical English capital.
+    pub cap_height: Option<f32>,
+    /// Distance from the baseline to the top of the lowercase "x" or
+    /// similar character.
+    pub x_height: Option<f32>,
+    /// Average width of all non-zero width characters in the font.
+    pub average_width: Option<f32>,
+    /// Maximum advance width of all characters in the font.
+    pub max_width: Option<f32>,
+    /// Metrics for an underline decoration.
+    pub underline: Option<Decoration>,
+    /// Metrics for a strikeout decoration.
+    pub strikeout: Option<Decoration>,
+    /// Union of minimum and maximum extents for all glyphs in the font.
+    pub bounds: Option<BoundingBox>,
+}
+
+impl Metrics {
+    /// Creates new metrics for the given font, size in pixels per em, and
+    /// normalized variation coordinates.
+    ///
+    /// If `size` is `None` or `Some(0.0)`, resulting metric values will be in font units.
+    pub fn from_font<'a>(
+        font: &impl TableProvider<'a>,
+        size: Option<f32>,
+        coords: &'a [NormalizedCoord],
+    ) -> Self {
+        let head = font.head();
+        let mut metrics = Metrics {
+            units_per_em: head.map(|head| head.units_per_em()).unwrap_or_default(),
+            ..Default::default()
+        };
+        let scale = compute_scale(size, metrics.units_per_em as f32);
+        if let Ok(head) = font.head() {
+            metrics.bounds = Some(BoundingBox {
+                x_min: head.x_min() as f32 * scale,
+                y_min: head.y_min() as f32 * scale,
+                x_max: head.x_max() as f32 * scale,
+                y_max: head.y_max() as f32 * scale,
+            });
+        }
+        if let Ok(maxp) = font.maxp() {
+            metrics.glyph_count = maxp.num_glyphs();
+        }
+        if let Ok(post) = font.post() {
+            metrics.is_monospace = post.is_fixed_pitch() != 0;
+            metrics.italic_angle = post.italic_angle().to_f64() as f32;
+            metrics.underline = Some(Decoration {
+                offset: post.underline_position().to_i16() as f32 * scale,
+                thickness: post.underline_thickness().to_i16() as f32 * scale,
+            });
+        }
+        let hhea = font.hhea();
+        if let Ok(hhea) = &hhea {
+            metrics.max_width = Some(hhea.advance_width_max().to_u16() as f32 * scale);
```

Comment:
Noob question, if you wrote scale * hhea.advance_width_max().to_u16() would you get to drop the cast?

Reply:
Rust doesn’t support implicit conversions regardless of operand order. This happens to be one of my favorite features of the language.